lib/Tools/LinearLayout.cpp

```diff
@@ -655,19 +655,95 @@ LinearLayout::divideRight(const LinearLayout &divisor) {
                          divisor.getInDimSizeLog2(inDim));
   }
 
+  // Record size 1 out-dims caused by the division.
   llvm::MapVector<StringAttr, int32_t> newOutDims = outDims;
-  for (const auto [outDim, outDimSize] : divisor.outDims) {
-    if (newOutDims[outDim] < outDimSize) {
+  llvm::DenseSet<size_t> sizeOneOutDimIndices;
+  for (auto [outDimName, outDimSize] : divisor.outDims) {
+    if (newOutDims[outDimName] < outDimSize) {
       return std::nullopt;
     }
-    newOutDims[outDim] /= outDimSize;
+    auto newOutDimSize = newOutDims[outDimName] / outDimSize;
+    if (newOutDimSize == 1) {
+      sizeOneOutDimIndices.insert(getOutDimIndex(outDimName));
+    }
+    newOutDims[outDimName] = newOutDimSize;
+  }
+
+  // Only erase the trailing empty out-dims.
+  // For example, consider l = o / r, where:
+  //
+  //   out-dims(o) = ["out0", "out1", "out2", "out3"]
+  //   out-dims(r) = ["out1", "out3"]
+  //
+  // If we remove "out1" from o, we get:
+  //
+  //   out-dims(l) = ["out0", "out2", "out3"]
```

Comment:
I am having trouble following this example.

It seems to be explaining an important edge case in an algorithm, but the overall algorithm -- the thing we're trying to do -- is not clear to me.  The example also has some assumptions that I don't understand.  For example, I don't understand how o / r returns anything at all if we remove "out1" from o.  Isn't that just an infeasible division?  (Unless you're assuming out1 is a size-zero dim?  But I don't see how I'm supposed to know that?)

I wonder if the following algorithm works.

1. Assume c = a*b.
2. We are given b and c, and we want to compute c.divRight(b).  i.e. we want to find `a` (or some a' which is equivalent to `a` ignoring size-zero dims).
3. Let b' be b but with all size-zero in-dims and out-dims removed.  Same for c'.
4. Compute our candidate quotient a' = c' / b', same as before.
5. Check if a' * b' == c', same as before.
6. If it matches, then let a'' be a' but with the minimum number of size-zero in- and out-dims added back to it so that a'' * b' = c as desired.
7. Return a''.

Reply:
About the new algorithm, I think it would fix the same problem as the existing code in this PR

Reply:
> For example, I don't understand how o / r returns anything at all if we remove "out1" from o. Isn't that just an infeasible division? (Unless you're assuming out1 is a size-zero dim? But I don't see how I'm supposed to know that?)

Yeah, we assume that "out1" is a size-zero dim. Why did you say "I don't see how I'm supposed to know that"? Is it because of lacking a comment, or you think there's no way to check it? If the former I'll add a comment; I thought it's clear because there's a variable `emptyOutDimIndices`.

Reply:
The key point is that we cannot remove arbitrary empty output dimensions from the quotient.

The following code simulates `quotient * divisor = result` and enumerates the output dimensions of the result from right to left to check which ones can be removed. When we perform the multiplication, the output dimensions of the quotient are always placed before the output dimensions of the divisor in the result. So if this order breaks in the result, we should stop enumerating output dimensions.

Therefore, I believe the new algorithm you described solves the same problem as the existing code.

Reply:
There are some confusion about `size-1` or `size-0` dimensions though. I call them `sizeOneOutDimIndices` since the out-dim maps everything to 0 and still has a size of 1.

Reply:
> Why did you say "I don't see how I'm supposed to know that"? Is it because of lacking a comment, or you think there's no way to check it?

Lacking a comment.  :)

> I thought it's clear because there's a variable emptyOutDimIndices.

...yeah I'm doing my best to understand what's going on, but I did not consider looking at the code below to understand the algorithm above (and anyway I'm not sure it would have helped me).

> There are some confusion about size-1 or size-0 dimensions though. I call them sizeOneOutDimIndices since the out-dim maps everything to 0 and still has a size of 1.

Ah yes, "size-1" dims is correct, I was calling it the wrong thing.

> Therefore, I believe the new algorithm you described solves the same problem as the existing code.

I think I would have an easier time understanding the new algorithm I proposed, but if you don't think that's the best approach (who knows, I'm not the one implementing it), it would help me if we could explain the algorithm we're using as a comment in the code.  I think you're starting to get there with this:

> The following code simulates quotient * divisor = result and enumerates the output dimensions of the result from right to left to check which ones can be removed. When we perform the multiplication, the output dimensions of the quotient are always placed before the output dimensions of the divisor in the result. So if this order breaks in the result, we should stop enumerating output dimensions.

I promise I'm doing my best to understand things here and not just being obstinate

Reply:
I've refactored the algorithm as follows:

1. Consider `c = a * b` and construct a candidate quotient `a'` first without removing any dimensions.
2. Check if `a' * b == c`. If yes, we start to remove empty dimensions.
3. First, we remove empty trailing output dimensions from `a'`.
4. Then, we remove empty trailing input dimensions from `a'`.
5. Finally, we return the final quotient `a''`.

Also, I found that we actually allow a linear layout with no input dimensions but with empty output dimensions. Therefore, the hacky `len(input_dims) == len(output_dims)` condition can be removed, allowing us to use the same code to check if data transfer can happen within each thread.

Reply:
Aren't empty dims inside `a'` also removable?  They're only not removable if `b` does not contain the dimension, right?